src/script/script.h

```diff
@@ -465,30 +492,16 @@ class CScript : public CScriptBase
 
     CScript& operator<<(const std::vector<unsigned char>& b) LIFETIMEBOUND
     {
-        if (b.size() < OP_PUSHDATA1)
-        {
-            insert(end(), (unsigned char)b.size());
-        }
-        else if (b.size() <= 0xff)
-        {
-            insert(end(), OP_PUSHDATA1);
-            insert(end(), (unsigned char)b.size());
-        }
-        else if (b.size() <= 0xffff)
-        {
-            insert(end(), OP_PUSHDATA2);
-            uint8_t _data[2];
-            WriteLE16(_data, b.size());
-            insert(end(), _data, _data + sizeof(_data));
-        }
-        else
-        {
-            insert(end(), OP_PUSHDATA4);
-            uint8_t _data[4];
-            WriteLE32(_data, b.size());
-            insert(end(), _data, _data + sizeof(_data));
-        }
-        insert(end(), b.begin(), b.end());
+        InsertSize(b.size());
+        InsertInput(b.begin(), b.end());
+        return *this;
+    }
+
+    template<size_t N>
+    CScript& operator<<(const std::array<unsigned char, N>& b) LIFETIMEBOUND
```

Comment:
> will this be touched again? 

Not sure, isn't that the case with vector as well?

> so it just seems like right now this is adding more code while leaving the underlying issue.

Can you please detail why you think this is the case? `_hex_v_u8` was mostly eliminated and the production code is basically the same (+3 lines for the other overload).

> Maybe this could be a span or a range?

I understood we wanted to avoid doing that in https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1692989490

Reply:
> however the change only fixes it for one very specific instance

Seems I missed some other cases, which other ones should we include here?

> leaving the problem otherwise unfixed

It does? This is what we've mentioned throughout the original PR by @hodlinator, @ryanofsky and yourself. I don't mind changing to span (it's what I suggested originally), if you think that would solve it in a cleaner way - but I need to understand the problem better since I though this fully solved our concerns regarding script data copying of hexadecimal values.

Reply:
I think this should just take a span, not an array or a vector because the container type should not be relevant.

The only reason to not use span here would be to try to make the cscript << operator act like the serialization << operator and treat vectors differently from spans and arrays. This behavior makes some sense for serialization, but doesn't make sense for pushing cscript data, so I don't think should be a concern.

(For serialization we use Span{} casts as a way of telling serialization API that data is fixed-length, not variable-length so should be formatted without a size prefix. But this is a dubious and confusing use of span. Seems like it would be better to have `FixedSize` and `VarSize` serialization formatters that would make code clearer and allow serializing variable size containers without size prefixes, and fixed size containers with them, since there are use-cases for both. Adding span or vector conversions just to get data serialized in the right format is unnecessarily confusing, and in some cases inefficient.)

Reply:
I'll investigate these tomorrow, thanks for the comments!

Reply:
> Would it make sense to drop operator<< from CScript, and instead have explicit "Append" and "Push" member functions (that take spans), and do the expected thing?

I think it makes sense, especially because cscripts are serializable, so it is confusing that `x << script` does an entirely different operation than `script << x`. But making this change would expand the scope of the PR beyond what it is currently doing which is just making cscript work a little better with hex literals from #30377.

On the idea of Append and Push methods, I think an Append method is not actually necessary, but may be useful. And if we think it is a good idea to disambiguate the CScript `<<`  operator, it probably also makes sense to disambiguate the CScript constructor, since it currently has one overload to add bytes to the script and other overloads to add opcodes and data. Probably a better interface would provide separate functions with descriptive names returning CScript objects, instead of providing multiple CScript constructors where you can't know what a constructor may be doing without knowing the type of the argument.

But I do think it would be nice if this PR just added support for std::span and std::byte without necessarily making bigger changes.

Reply:
Having an elegant way of expressing scripts in code is nice. How about (omitted some hex digits):
```C++
script << "046711d5f"_sized_hex << OP_CHECKSIG;
```
`_sized_hex` could possibly create something like a `SizePrefixed(std::array<std::byte>)` object and `CScript` could have a `<<` operator taking such objects and prepending the size.

> But I do think it would be nice if this PR just added support for std::span and std::byte without necessarily making bigger changes.

That was what I was originally doing as part of #30377, I dropped it so that this PR here could let us work toward some final form of this aspect. So I would rather see a bigger step being made.

Reply:
> But I do think it would be nice if this PR just added support for std::span and std::byte without necessarily making bigger changes.

I've added support for `const std::span<const unsigned char>` (for vector & array), but `std::byte` seems like a bigger change (as explained now in the PR's description) - unless you'd like me to add it as an extra overload - let me know if that's what you meant.

Reply:
> An overload for `std::byte` could be one more line of code, but would avoid having to touch all the touched non-header lines in this pull request again.

Yeah sorry for any confusion. My suggestion in https://github.com/bitcoin/bitcoin/pull/30765#discussion_r1741079145 here was to use `std::byte` if it was easy or `BasicByte` as a fallback if would be too much work to convert other code to use std::byte. Having two overloads is also fine.

Reply:
I have [added](https://github.com/bitcoin/bitcoin/compare/ab875e14b7ed02e6e4bdc607d67e28bc10216a27..14080bb53d649ed1c99f121f377d2c8f2019c99e) an `std::byte` overload in a separate commit, so now we have:
```C++
CScript& operator<<(const std::span<const std::byte> b) LIFETIMEBOUND
{
    return *this << std::bit_cast<std::span<uint8_t>>(b);
}
```
which delegates to:
```C++
CScript& operator<<(const std::span<const unsigned char> b) LIFETIMEBOUND
{
    AppendDataSize(b.size());
    AppendData(b.begin(), b.end());
    return *this;
}
```

I haven't tried `BasicByte`, but this seems to work - if the other compilers agree...